mutable_tree.go

```diff
@@ -0,0 +1,446 @@
+	return latestVersion, nil
+}
+
+// GetImmutable loads an ImmutableTree at a given version for querying
+func (tree *MutableTree) GetImmutable(version int64) (*ImmutableTree, error) {
+	rootHash := tree.ndb.getRoot(version)
+	if rootHash == nil {
+		return nil, ErrVersionDoesNotExist
+	} else if len(rootHash) == 0 {
+		return &ImmutableTree{
+			ndb:     tree.ndb,
+			version: version,
+		}, nil
+	}
+	return &ImmutableTree{
+		root:    tree.ndb.GetNode(rootHash),
+		ndb:     tree.ndb,
+		version: version,
+	}, nil
+}
+
+// Rollback resets the working tree to the latest saved version, discarding
+// any unsaved modifications.
+func (tree *MutableTree) Rollback() {
+	if tree.version > 0 {
+		tree.ImmutableTree = tree.lastSaved.clone()
+	} else {
+		tree.ImmutableTree = &ImmutableTree{ndb: tree.ndb, version: 0}
+	}
+	tree.orphans = map[string]int64{}
+}
+
+// GetVersioned gets the value at the specified key and version.
+func (tree *MutableTree) GetVersioned(key []byte, version int64) (
+	index int, value []byte,
+) {
+	if tree.versions[version] {
+		t, err := tree.GetImmutable(version)
+		if err != nil {
+			return -1, nil
+		}
+		return t.Get(key)
+	}
+	return -1, nil
+}
+
+// SaveVersion saves a new tree version to disk, based on the current state of
+// the tree. Returns the hash and new version number.
+func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
+	version := tree.version + 1
+
+	if tree.root == nil {
+		// There can still be orphans, for example if the root is the node being
+		// removed.
+		debug("SAVE EMPTY TREE %v\n", version)
+		tree.ndb.SaveOrphans(version, tree.orphans)
+		tree.ndb.SaveEmptyRoot(version)
+	} else {
+		debug("SAVE TREE %v\n", version)
+		// Save the current tree.
+		tree.ndb.SaveBranch(tree.root)
+		tree.ndb.SaveOrphans(version, tree.orphans)
+		tree.ndb.SaveRoot(tree.root, version)
+	}
+	tree.ndb.Commit()
+	tree.version = version
+	tree.versions[version] = true
+
+	// Set new working tree.
+	tree.ImmutableTree = tree.ImmutableTree.clone()
+	tree.lastSaved = tree.ImmutableTree.clone()
+	tree.orphans = map[string]int64{}
+
+	return tree.Hash(), version, nil
+}
+
+// DeleteVersion deletes a tree version from disk. The version can then no
+// longer be accessed.
+func (tree *MutableTree) DeleteVersion(version int64) error {
+	if version == 0 {
+		return cmn.NewError("version must be greater than 0")
+	}
+	if version == tree.version {
+		return cmn.NewError("cannot delete latest saved version (%d)", version)
+	}
+	if _, ok := tree.versions[version]; !ok {
+		return cmn.ErrorWrap(ErrVersionDoesNotExist, "")
+	}
+
+	tree.ndb.DeleteVersion(version)
+	tree.ndb.Commit()
+
+	delete(tree.versions, version)
+
+	return nil
+}
+
+// Rotate right and return the new node and orphan.
+func (tree *MutableTree) rotateRight(node *Node) (newNode *Node, orphan *Node) {
+	version := tree.version + 1
+
+	// TODO: optimize balance & rotate.
+	node = node.clone(version)
+	l := node.getLeftNode()
+	_l := l.clone(version)
```

Comment:
Can we rename `_l` to `newNode` and `l` to `orphaned`? Maybe we can remove the named returns here, too (`func (tree *MutableTree) rotateRight(node *Node) (*Node, *Node)` instead)